### DISCUSSION TITLE
Custom matching function on routes

### DISCUSSION THREAD

**Discussion by modi98** [remix-run/react-router#11546] [Proposals]
I have a web application that has unconventional segments which used to work on react router 5 with the use of multiple slugs and validators for path matching, however with the new path matching from v6 it got a bit harder to match this controllers.

Here's a few examples:
www.my-site.com/calendar.event1.create
www.my-site.com/calendar.event2.edit

I know I can just change it to use `/` but unfortunately that would be a breaking change that we're trying to avoid, and also this is a very simple example, there are more complex ones that are not easily solvable. **I'll be willing to contribute this feature** but would love to get some feedback instead of just wasting my time doing something that won't ever get merged.

End product would look something like this:
```jsx
<BrowserRouter>
  <Routes customMatcher={myMatchingFunction}>
    <Route path="/calendar.{eventId}.create" element={<CreateEvent />} />
    <Route path="/calendar.{eventId}.edit" element={<EditEvent />} />
  </Routes>
</BrowserRouter>
```

**Comment by kiliman** [comment]
I think the best way to implement something like this is if React Router used a tree-matching algorithm.

NOTE: The link below is from Remix, but since it uses React Router, it should still apply.

https://github.com/remix-run/remix/issues/4857

**Reply by modi98** [reply]
I'm not aware of the innings of react-router's route matching, I thought the `Routes` component just used `matchPath` internally to render whatever matches, could it be as simple as allowing the users to override this function with their `customMatchPath`?

**Comment by modi98** [comment]
I opened up https://github.com/remix-run/react-router/pull/11556 as a POC for this feature
